docs/reference/associations.txt

```diff
@@ -1255,30 +1255,101 @@ The child may become orphaned if it is ordinarily only referenced via the parent
 
    Band.first.delete # Will delete all associated albums.
 
+
 Autosaving
 ----------
 
-One core difference between Mongoid and Active Record from a behavior standpoint
-is that Mongoid does not automatically save associated documents for
-non-embedded associations. This is for performance reasons.
+One core difference between Mongoid and ActiveRecord is that Mongoid does not
+automatically save associated documents for referenced (i.e., non-embedded)
+associations when the parent is saved, for performance reasons.
 
-To enable an autosave on a non-embedded association (embedded associations do not need
-this since they are actually part of the parent in the database) add the autosave
-option to the association.
+If autosaving is not used, it is possible to create dangling references
+to non-existent documents via associations:
 
-Note that autosave functionality will automatically be added to an association when using
-``accepts_nested_attributes_for`` or validating presence of the association.
+.. code-block:: ruby
+
+  class Band
+    include Mongoid::Document
+    
+    has_many :albums
+  end
+
+  class Album
+    include Mongoid::Document
+    
+    belongs_to :band
+  end
+  
+  band = Band.new
+  album = Album.create!(band: band)
+  
+  # The band is not persisted at this point.
+  
+  album.reload
+  
+  album.band_id
+  # => BSON::ObjectId('6257699753aefe153121a3d5')
+  
+  # Band does not exist.
+  album.band
+  # => nil
+
+To make referenced associations save automatically when the parent is saved,
+add the ``:autosave`` option to the association:
 
 .. code-block:: ruby
 
-   class Band
-     include Mongoid::Document
-     has_many :albums, autosave: true
-   end
+  class Band
+    include Mongoid::Document
+    
+    has_many :albums
+  end
+
+  class Album
+    include Mongoid::Document
+    
+    belongs_to :band, autosave: true
+  end
+  
+  band = Band.new
+  album = Album.create!(band: band)
+  
+  # The band is persisted at this point.
+  
+  album.reload
+  
+  album.band_id
+  # => BSON::ObjectId('62576b4b53aefe178b65b8e3')
+  
+  album.band
+  # => #<Band _id: 62576b4b53aefe178b65b8e3, >
+
+The autosaving functionality is automatically added to an association when
+using ``accepts_nested_attributes_for``, so that the application does not
+need to track which associations were modified when processing a form
+submission.
+
+Embedded associations always autosave, because they are stored as part of the
+parent document.
+
+Some operations on associations always save the parent and the child documents
+as part of the operation, regardless of whether autosaving is enabled.
+A non-exhaustive list of these operations is as follows:
+
+- Assignment to the association:
+
+  .. code-block:: ruby
+  
+    # Saves the band and the album.
+    band.albums = [Album.new]
+
+- ``push``, ``<<``:
+
+  .. code-block:: ruby
+  
+    band.albums << Album.new
+    band.albums.push(Album.new)
```

Comment:
Is there a reason why these methods don't follow autosave? I personally think this is pretty confusing, however, given the apps relying on this behavior, it might be hard to change.

Perhaps we should clarify that the workaround to not use autosave is to do:

```ruby
new_album = Album.new
band.album_ids << new_album._id
new_album.save!
```

I tend to do this workaround a lot in my apps.

Reply:
> Is there a reason why these methods don't honor the autosave setting? 

Don't know.

> Perhaps we should clarify that the workaround to not use autosave is to do:

You are welcome to use it but because you no longer have references to objects in the tree the object tree is essentially broken and for this reason I personally wouldn't put it in official documentation as a recommendation.

> the current behavior uses atomic push method, which doesn't call callbacks on the band in the above example?

If you think this is a bug please feel free to create a ticket.

Reply:
> atomic push... If you think this is a bug please feel free to create a ticket.

Not a "bug", but rather your documentation in this PR uses the word "save" which is misleading if it's actually an atomic push. "save" implies that callbacks are fired, as per the `.save` method behavior.

Reply:
> the object tree is essentially broken

Hmmm... this comment reveals an important philosophical difference between "Mongoid in theory" and "Mongoid in the real world". The object tree **will always be broken**, whether you want it to be or not. Because errors can happen at anytime (i.e. mid-controller action), you will always have some hanging document references.

Moreover, as per my comments below, autosave exacerbates the problem because you end up doing a document save in a partial state, i.e. the relation you are adding, but not the attributes. It's far safer to commit the entire document (all relations + all attributes) at once, otherwise you end up with partially-saved documents.

Reply:
I don't follow. Even if there are some other issues with referential integrity/unspecified errors (?) this isn't a reason to intentionally break object tree.

Reply:
@johnnyshields if you want to create an object for an association without it being persisted to the database, you can use the `build*` methods:
```
For an _many relationship:
User.accounts.build

For a _one relationship:
User.build_pet
```

Reply:
@Neilshweky my core point here is I don't understand why:

```ruby
# This doesn't persist
User.accounts.build

# This does persist
account = Account.new
User.accounts << account
```

Most users will find this highly counter intuitive.

Reply:
@p-mongo my point is that implementers should be responsible for persisting their own objects (with `save!` method) and maintaining the object tree. Mongoid should not try to "enforce" the object tree by persisting when the user doesn't expect it.